Create email list page for authflow v2

Kuixz opened the pull request:
It seems this one was put on top of the birthdate PR (#5). Oh no. :(

Don't remember why. I suspect it is because of a styling class originally in the birthdate PR, or something like that.

Kuixz committed:
Create Go handler for settings/identity/email

Don't use the wrong verification service

Enforce most-recently edited order

Sort by time directly, remove comment and extract route

Fix typo (Requires route)

Kuixz committed:
Create HTML/CSS for settings/identity/email

Fix translation key names

Add arrows to view page when multiple emails

Correct back arrow redirect

Fix CSS

Use rem in css

Make css more targeted

Fix CSS

Remove incorrect sort and verify buttons

Use page-navbar to render navbar

Cherry-pick edits from view and use ol

Fix HTML

yyorkchan committed:
Route to settings change password page

yyorkchan committed:
Introduce account management service for non-sdk change password

yyorkchan committed:
Refactor password error viewmodel for usability

yyorkchan committed:
Introduce settings change password page

Kuixz commented:
This PR is no longer on top of the birthday one. :)

yyorkchan committed:
Support sdk settings change password

yyorkchan committed:
Make generate after rebase

yyorkchan committed:
Fix err check ci

yyorkchan committed:
Update navbar

cychiuae commented on authui/src/authflowv2/components/settings-identity-email.css:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  .settings-identity-email-indent::before {
+    @apply w-full h-full flex justify-center;
+    line-height: 24px;
+    font-size: 21px;
+    font-family: "material-icons";
+    color: var(--settings-identity-email-text-color);
+
+    content: "";
```
Remove these styles to `settings-identity-email-indent.with-icon::before`

cychiuae commented on authui/src/authflowv2/components/settings-identity-email.css:
```diff
@@ -0,0 +1,96 @@
+
+  .settings-identity-email-indent::before {
+    @apply w-full h-full flex justify-center;
+    line-height: 24px;
+    font-size: 21px;
+    font-family: "material-icons";
```
I find that we can just use `@apply material-icons` to replace these three lines

Kuixz replied:
it seems to me that `material-icons` css class in `material-symbols.css` doesn't have line height & font size declaration? 🤔

cychiuae replied:
oh. You are right. I think you can keep line-height and font-size and then remove font-family

cychiuae commented on authui/src/authflowv2/components/settings-identity-email.css:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  .settings-identity-email-indent {
+    @apply grid grid-cols-[24px_1fr_auto] gap-2;
```
We use `rem` instead of `px`

https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units

cychiuae commented on pkg/auth/handler/webapp/authflowv2/settings_identity_email.go:
```diff
@@ -0,0 +1,132 @@
+func ConfigureSettingsIdentityEmailRoute(route httproute.Route) httproute.Route {
+	return route.
+		WithMethods("OPTIONS", "POST", "GET").
+		WithPathPattern("/settings/identity/email")
```
Extract this constants and define it in `routes.go`

cychiuae commented on pkg/auth/handler/webapp/authflowv2/settings_identity_email.go:
```diff
@@ -0,0 +1,132 @@
+				Email:      claim.Value,
+				Verified:   claim.Verified,
+				Verifiable: claim.EndUserTriggerable,
+				UpdatedAt:  id.UpdatedAt.Unix(),
```
I prefer keep UpdatedAt to be `time.Time` coz we have template function to format date

cychiuae commented on pkg/auth/handler/webapp/authflowv2/settings_identity_email.go:
```diff
@@ -0,0 +1,132 @@
+			continue
+		}
+
+		claim := claims[0]
```
What if there are more than 1 claim. How would we handle it?

Kuixz replied:
I wouldn't know :doglaugh: I just reuse the logic I see in the authflowv1. Maybe it's stored in reverse chronological order?

cychiuae commented on pkg/auth/handler/webapp/authflowv2/settings_identity_email.go:
```diff
@@ -0,0 +1,132 @@
+	})
+
+	ctrl.PostAction("verify", func() error {
+		// opts := webapp.SessionOptions{
+		// 	RedirectURI:     redirectURI,
+		// 	KeepAfterFinish: true,
+		// }
+		// intent := intents.NewIntentVerifyIdentity(userID, model.IdentityTypeLoginID, identityID)
+
+		// result, err := ctrl.EntryPointPost(opts, intent, func() (input interface{}, err error) {
+		// 	input = nil
+		// 	return
+		// })
+		// if err != nil {
+		// 	return err
+		// }
```
Remove these comments.

cychiuae commented on resources/authgear/templates/en/web/authflowv2/settings_identity_email.html:
```diff
@@ -0,0 +1,89 @@
+<div class="mt-8 settings-content flex flex-col gap-5">
+    {{ $hasMultiple := gt (len .CheckedEmails) 1 }}
+
+    <span>
```
Use `ul` instead of `span`

cychiuae commented on resources/authgear/templates/en/web/authflowv2/settings_identity_email.html:
```diff
@@ -0,0 +1,89 @@
+
+    <span>
+        {{ range $index, $element := .CheckedEmails }}
+            <div class="settings-identity-email-block">
```
It should be `li`

cychiuae commented on resources/authgear/templates/en/web/authflowv2/settings_identity_email.html:
```diff
@@ -0,0 +1,89 @@
+
+{{ define "page-content" }}
+
+<div class="mt-8 settings-content flex flex-col gap-5">
```
Please remove `mt-8`

cychiuae commented on resources/authgear/templates/en/web/authflowv2/settings_identity_email.html:
```diff
@@ -0,0 +1,89 @@
+{{ define "page-navbar" }}
+    {{ template "authflowv2/__navbar.html"
+    (dict
+    "BackHref" (call $.MakeURL "/settings")
```
Missing "BackTitle"

cychiuae commented on resources/authgear/templates/en/web/authflowv2/settings_identity_email.html:
```diff
@@ -0,0 +1,89 @@
+                            class="secondary-btn"
+                            type="submit"
+                            name="verify"
+                            {{/* value="{{ $element.identity_id }}" */}}
+                            >{{ translate "v2.page.settings-identity-email.default.verify-label" . }}
+                        </button></span>
+                    </div>
+                {{ end }}
+            </div>
+        {{ end }}
+    
+        <div class="py-6 flex justify-center body-text--lg">
+            <a
+                href="/settings/identity/email/new">
+                {{ translate "v2.page.settings-identity-email.default.add-new-label" . }}
+            </a>
+        </div>
+    </span>
+    {{ if $hasMultiple }}
+    <span>
+        <div class="settings-identity-email-subtitle">
+            {{ translate "v2.page.settings-identity-email.default.primary-email-label" . }}
+        </div>
+        <div class="settings-identity-email-block">
+            <div class="settings-identity-email-indent with-icon">
+                <div class="settings-identity-email-text">{{ .Email }}</div>
+                <span class="body-text--lg"><a 
+                href="/settings/identity/email/change">
+                    {{ translate "v2.page.settings-identity.default.change-label" . }}
+                </a></span>
+            </div>
+            <div class="settings-identity-email-indent pt-1">
+                <div class="settings-identity-email-subtext success">
+                {{ translate "v2.page.settings-identity-email.default.verified-label" . }}
+                </div>
+            </div>
+        </div>
+    </span>
+    {{ end }}
+</div>
```
Please amend the layout according to the latest design

cychiuae commented on resources/authgear/templates/en/web/authflowv2/settings_identity_email.html:
```diff
@@ -0,0 +1,89 @@
+                        arrow_forward_ios
+                    </a>
+                    {{ else }}
+                    <span class="body-text--lg"><a 
+                        href="/settings/identity/email/edit">
+                        {{ translate "v2.page.settings-identity.default.edit-label" . }}
+                    </a></span>
```
I prefer you write this in this format
```
<span>
  <a />
</span>
```

It would be more readable

cychiuae commented on resources/authgear/templates/en/web/authflowv2/settings_identity_email.html:
```diff
@@ -0,0 +1,89 @@
+                    {{ end }}
+                </div>
+                <div class="settings-identity-email-indent pt-1">
+                    {{ if $element.Verified }}
+                    <div class="settings-identity-email-subtext success">
+                    {{ translate "v2.page.settings-identity-email.default.verified-label" . }}
+                    {{ else }}
+                    <div class="settings-identity-email-subtext failure">
+                    {{ translate "v2.page.settings-identity-email.default.not-verified-label" . }}
+                    {{ end }}
+                    </div>
```
All the if-branches should be a complete expression.
i.e. the close tag of the div should be place in the branch

```html
{{ if some_condition }}
  <div></div>
{{ else }}
  <div></div>
{{ end }}
```

Don't be lazy

cychiuae reviewed (commented):
Please configure your editor to use 2 spaces for indentation in the html file

cychiuae commented:
<img width="2184" alt="Screenshot 2024-09-05 at 3 39 01 PM" src="https://github.com/user-attachments/assets/8a884b7b-5706-4afc-9d92-83954bce8c03">

cychiuae reviewed (changes requested):
Please amend the target url of email item in `authflowv2/settings.html`. It should navigate to the page you are working on

cychiuae commented:
<img width="1244" alt="Screenshot 2024-09-05 at 3 46 28 PM" src="https://github.com/user-attachments/assets/1674be8a-ff15-45cf-bb68-881cbcf648c8">

Please amend the styles so that content is visible in dark mode

Kuixz committed:
Link to email page in settings

Kuixz commented:
Discuss offline, restated here:

Cyka blyat forgot to apply the css class I already defined.

Kuixz committed:
Fix missing translation

yyorkchan committed:
Rename variables

yyorkchan committed:
Refactor settings change password style to use settings-content

yyorkchan committed:
Make fmt

yyorkchan committed:
Wrap with s.Database.WithTx & output format

cychiuae committed:
Rename Controller.Serve -> Controller.ServeWithDBTx

I want developers to explicitly state that
they want the controller serve with db tx

cychiuae committed:
Add back controller.Serve

But now it doesn't open database transaction by default

cychiuae committed:
Generate translation

cychiuae committed:
Use proper case for a word

cychiuae commented:
@Kuixz Can you rebase?

Kuixz commented:
Completed rebase @cychiuae, i think everything is intact. Will rebase 11 and 15 shortly

yyorkchan committed:
Wrap Change password with Database.WithTx

cychiuae committed:
Explicitly open connection for change password
